### PR TITLE
Feature/vitejs svelte

### PR DESCRIPTION
**Pull request by gwendalF**
- #26

### CODE REVIEW

**Comment by kirillt** [comment]
Bugs:
- [ ] When a link is pasted, but title and description are not loaded yet:
  clicking on **"AUTO FILLED" causes "Fill out this field" message**.
- [ ] When a link is pasted, and title and description are already loaded:
  clicking on **"AUTO FILLED" attempts to store the link**.
- [ ] If the link is already saved, and title and description are already loaded:
    clicking on **"AUTO FILLED" displays "Link created" message**.
- [ ] If the link is already saved, and title and description are already loaded:
    clicking on **"CREATE" displays "Link created" message**.

- [ ] Clicking on **"DELETE" button panics** with:
  ```
  cannot remove the link: Os {
    code: 2, kind: NotFound, message: "No such file or directory"
  }
  ```

Correct:
- Links are created in `~/.ark-shelf` and restarting the app after deleting a link file from the folder leads to this link disappearing.
- Frames with description popup by hovering as expected.

Fine for now:
- The description is duplicated between list items and popups.
- Scores don't work.

**Comment by gwendalF** [comment]
The errors delete panics is because the file is already deleted. It's still see because I have a duplicate key

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -1,45 +1,48 @@
...
+                    pkg-config
+
+            - name: Fetch Node.js Dependencies
+              uses: pnpm/action-setup@v2
+              with:
+                  version: 8
+                  run_install: true
+
+            - name: Build Release version
+              run: pnpm build
```
Comment (mberry) [inline .github/workflows/release.yml:42]: This should be `pnpm tauri build`

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -1,45 +1,48 @@
...
+
+            - name: Release AppImage
+              uses: ncipollo/release-action@v1
+              with:
+                  artifacts: './target/release/bundle/appimage/*.AppImage'
```
Comment (mberry) [inline .github/workflows/release.yml:47]: Should be `'./src-tauri/target/release/bundle/appimage/*.AppImage'`

File: dist/index.html
Comment (mberry) [inline dist/index.html:1]: Should dist be checked in? These are build files. 
Whole folder needs to go in .gitignore

**Review by mberry — CHANGES_REQUESTED** [review]
* `release.yaml` needs a fix
* `./dist` in .gitignore

* The dates are wrong now, not sure what commit caused it, but clearly something happening here:
Screenshot_20230919_084340

* Sort options only work ascending, if you click it again nothing changes.

* Alphabetically sorting isn't case normalised, so Xero is before alpha
 
 * "There is already a link with the same URL" does a big redraw of the form at smaller widths, text could just wrap around and not touch the box size. When you go to type in a new link it redraws again.

**Comment by gwendalF** [comment]
The sorting is workign as before. I corrected it to use lowercase. If we want to make it works descending I think another PR is better. The date is fixed as the release and dist folder

**Comment by gwendalF** [comment]
The times taken is from the write_to_path at the arkilb repo: 
image
This function computes the preview (Why I don't know) and fetch the images. I can make a really faster function that just write_to_path (No pun intended)

 - [x] Pasting new link should erase title and description fields
 - [ ] Sometimes Error deleting the link on just created link -> Need more investigation but this function just tries to delete the files. It may be the wrong path it there were an issue on the frontend during the creation

**Comment by kirillt** [comment]
@gwendalF 
Pasting here the function just for the context..
```
 51     /// Write zipped file to path                                              // the comment is misleading now
 52     pub async fn write_to_path<P: AsRef<Path>>(
 53         &mut self,
 54         root: P,
 55         path: P,
 56         save_preview: bool,
 57     ) -> Result<(), Error> {
 58         let id = self.id()?;
 59         store_meta::<Metadata, _>(root.as_ref(), id, &self.meta)?;
 60 
 61         let mut link_file = File::create(path.as_ref().to_owned())?;
 62         let file_data = self.url.as_str().as_bytes();
 63         link_file.write(file_data)?;
 64         if save_preview {
 65             let preview_data = Link::get_preview(self.url.clone())
 66                 .await
 67                 .unwrap_or_default();
 68             let image_data = preview_data
 69                 .fetch_image()
 70                 .await
 71                 .unwrap_or_default();
 72             self.save_preview(root.as_ref(), image_data, id)
 73                 .await?;
 74         }
 75 
 76         store_meta::<Metadata, _>(root, id, &self.meta)
 77     }
```

Yes, this function does 3 things alltogether and in fact they can be done in parallel. Feel free to change this in arklib, or write your own function in Shelf, and move it to arklib after we test it.

**Comment by mberry** [comment]
The original commit is quite a good base, could start from there and draw in?

File: src-tauri/src/command/mod.rs [hunk 3]
```diff
@@ -28,31 +28,46 @@ pub struct LinkScoreMap {
 /// Create a `.link`
 async fn create_link(
     url: String,
-    state: tauri::State<'_, Cli>,
+    meta_data: arklib::link::Metadata
 ) -> Result<String> {
     let url = Url::parse(url.as_str())?;
     let resource = arklib::id::ResourceId::compute_bytes(url.as_ref().as_bytes())
```
Comment (kirillt) [inline src-tauri/src/command/mod.rs:34]: Nitpick: just `id` is fine, `resource` is more complex structure

File: src-tauri/src/command/mod.rs [hunk 4]
```diff
@@ -28,31 +28,46 @@ pub struct LinkScoreMap {
...
 }
 
 #[tauri::command]
 /// Remove a `.link` from directory
-async fn delete_link(name: String, state: tauri::State<'_, Cli>) -> Result<()> {
-    fs::remove_file(format!("{}/{}", &state.path, name))?;
+async fn delete_link(name: String) -> Result<()> {
+    let path = ARK_SHELF_WORKING_DIR.get().unwrap().join(&name);
+    fs::remove_file(&path)?;
+    let resource = name.rsplit("-").enumerate().take_while(|(u,_)| {
```
Comment (kirillt) [inline src-tauri/src/command/mod.rs:55]: This looks fragile.. We should construct `id` only from the content, i.e. from the link itself (URL / file content).

File: src-tauri/src/main.rs [hunk 5]
```diff
@@ -62,14 +60,36 @@ struct PreviewLoaded {
     created_time: Option<SystemTime>
 }
 
-async fn get_preview(path: &std::path::PathBuf, manager: AppHandle) -> Result<()> {
+fn init_statics_and_dir() {
+    let cli = Cli::parse();
+    let working_dir = PathBuf::from(&cli.path);
+    ARK_SHELF_WORKING_DIR.set(working_dir).unwrap();
+    let scores_path = PathBuf::from(&cli.path).join(arklib::STORAGES_FOLDER).join("shelf").join("scores");
```
Comment (kirillt) [inline src-tauri/src/main.rs:67]: Just `.join(arklib::STORAGES_FOLDER).join("scores");` without `.join("shelf")`. Storages should be usable across the apps.

File: src-tauri/src/command/mod.rs [hunk 6]
```diff
@@ -51,16 +49,13 @@ async fn create_link(
 /// Remove a `.link` from directory
 async fn delete_link(name: String) -> Result<()> {
     let path = ARK_SHELF_WORKING_DIR.get().unwrap().join(&name);
+    let content = fs::read_to_string(&path)?;
+    let id = arklib::id::ResourceId::compute_bytes(content.as_bytes())
```
Comment (kirillt) [inline src-tauri/src/command/mod.rs:53]: Yep :+1: Later we'll integrate with the index from `arklib` so we won't recompute the ids all the time.

Comment (gwendalF) [reply src-tauri/src/command/mod.rs:53]: the name is currently: domain-id. That's why I used to split with the - to get the id without rereading the file before

Comment (kirillt) [reply src-tauri/src/command/mod.rs:53]: Yeah, I understand. Still, it's less direct than counting id from the content itself, and also URLs are small enough to compute their id quickly.

**Comment by kirillt** [comment]
All my previous comments are fixed :+1: 

Only one new issue:
- [ ] Starting the app with non-empty working dir (when there are some `.link` files) crashes:
```
[src/command/mod.rs:85] &item = DirEntry("/home/kirill/.ark-shelf/github.com-57-731385041.link")
[src/command/mod.rs:89] &path_list = [
    "github.com-57-731385041.link",
]
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: No such file or directory (os error 2)', src/command/mod.rs:186:77
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panicking.rs:67:14
   2: core::result::unwrap_failed
             at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/result.rs:1651:5
   3: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
   4: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
   5: tauri::hooks::InvokeResolver<R>::respond_async_serialized::{{closure}}
   6: tokio::runtime::task::core::Core<T,S>::poll
   7: tokio::runtime::task::harness::Harness<T,S>::poll
   8: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
   9: tokio::runtime::scheduler::multi_thread::worker::Context::run
  10: tokio::runtime::context::scoped::Scoped<T>::set
  11: tokio::runtime::context::runtime::enter_runtime
  12: tokio::runtime::scheduler::multi_thread::worker::run
  13: tokio::runtime::task::core::Core<T,S>::poll
  14: tokio::runtime::task::harness::Harness<T,S>::poll
  15: tokio::runtime::blocking::pool::Inner::run
```
- [ ] Also, please update GitHub Actions like here:
https://github.com/ARK-Builders/arklib/commit/2d71b8aca5059aad05c339ff115bf30b116375b8

File: src-tauri/src/base/mod.rs [hunk 7]
```diff
@@ -1,8 +1,9 @@
...
+use std::path::Path;
 
-pub use arklib::link::{Link, OpenGraph};
+pub use arklib::link::{Link, Metadata, OpenGraph};
 use serde::{Deserialize, Serialize};
 use walkdir::{DirEntry, WalkDir};
+
 #[derive(Debug, Serialize, Deserialize)]
 pub struct LinkScoreMap {
     pub name: String,
```
Comment (kirillt) [inline src-tauri/src/base/mod.rs:9]: Do we really need `name`? Can we use `id` instead?

File: src-tauri/src/base/mod.rs [hunk 8]
```diff
@@ -27,30 +28,45 @@ pub enum Mode {
     Score,
 }
 pub type Scores = Vec<Score>;
+
 #[derive(Debug, Deserialize, Serialize, PartialEq, PartialOrd, Clone)]
 pub struct Score {
     pub name: String,
```
Comment (kirillt) [inline src-tauri/src/base/mod.rs:34]: Do we really need `name` here?

File: src-tauri/src/base/mod.rs [hunk 9]
```diff
@@ -27,30 +28,45 @@ pub enum Mode {
...
 
-    pub hash: String,
+    pub id: String,
```
Comment (kirillt) [inline src-tauri/src/base/mod.rs:36]: Can we use `ResourceId` directly from arklib here?

```
#[derive(
    Eq,
    Ord,
    PartialEq,
    PartialOrd,
    Hash,
    Clone,
    Copy,
    Debug,
    Deserialize,
    Serialize,
)]
pub struct ResourceId {
    pub data_size: u64,
    pub crc32: u32,
}
```

File: src-tauri/src/base/mod.rs [hunk 10]
```diff
@@ -27,30 +28,45 @@ pub enum Mode {
...
-            "{:x}",
-            arklib::id::ResourceId::compute(
-                File::open(&path).unwrap().metadata().unwrap().len(),
-                path,
-            )
-            .crc32
-        )
+    pub fn new(url: &str) -> Self {
+        let id = arklib::id::ResourceId::compute_bytes(url.as_bytes()).unwrap();
+        let id = format!("{}-{}", id.crc32, id.data_size);
```
Comment (kirillt) [inline src-tauri/src/base/mod.rs:45]:
```
impl Display for ResourceId {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}-{}", self.data_size, self.crc32)
    }   
}
```

File: src-tauri/src/command/mod.rs [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    ARK_SHELF_WORKING_DIR, METADATA_PATH, PREVIEWS_PATH, SCORES_PATH,
+};
+
+use tauri::{Builder, Runtime};
+use url::Url;
+use walkdir::{DirEntry, WalkDir};
+
+#[derive(serde::Serialize)]
+pub struct LinkScoreMap {
+    pub name: String,
```
Comment (kirillt) [inline src-tauri/src/command/mod.rs:23]: Do we really need `name` here?

File: src-tauri/src/command/mod.rs [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+    pub id: String,
```
Comment (kirillt) [inline src-tauri/src/command/mod.rs:24]: Can we use `ResourceId` directly here?

**Comment by kirillt** [comment]
Works good, only 1 minor UI issue:
- [ ] When a link is added, we should erase input fields

**Comment by gwendalF** [comment]
There is still a small issue concerning description. But I think we can resolve it with the issue https://github.com/ARK-Builders/ARK-Shelf-Desktop/issues/45

**Review by kirillt — APPROVED** [review]
Good job, thanks

**Pull request merged**